packages/dds/tree/src/test/simple-tree/simpleTree.bench.ts

```diff
@@ -0,0 +1,190 @@
+			benchmark({
+				type: benchmarkType,
+				title: `Wide Tree as SimpleTree: reads with ${numberOfNodes} nodes`,
+				before: () => {
+					// Setup
+					tree = generateWideSimpleTree(numberOfNodes, leafValue);
+					for (let i = 0; i < numberOfNodes; i++) {
+						expected += leafValue;
+					}
+				},
+				benchmarkFn: () => {
+					const { nodesCount, sum } = readWideSimpleTree(tree);
+					actualNodesCount = nodesCount;
+					actualSum = sum;
+				},
+				after() {
+					assert.equal(actualNodesCount, numberOfNodes);
+					assert.equal(actualSum, expected);
+				},
+			});
+		}
+	});
+
+	describe(`Edit SimpleTree`, () => {
+		const leafValue = 1;
+		const changedLeafValue = -1;
+		for (const [numberOfNodes, benchmarkType] of nodesCountDeep) {
+			let tree: DeepTreeNode;
+			benchmark({
+				type: benchmarkType,
+				title: `Update value at leaf of ${numberOfNodes} deep tree`,
+				before: () => {
+					// Setup
+					tree = generateDeepSimpleTree(numberOfNodes, leafValue);
+				},
+				benchmarkFn: () => {
+					writeDeepSimpleTree(tree, changedLeafValue);
+				},
+				after: () => {
+					const expected = generateDeepSimpleTree(numberOfNodes, changedLeafValue);
+					assert.deepEqual(tree, expected);
+				},
+			});
+		}
+
+		for (const [numberOfNodes, benchmarkType] of nodesCountWide) {
+			let tree: WideTreeNode;
+			benchmark({
+				type: benchmarkType,
+				title: `Remove and insert end value at leaf of ${numberOfNodes} Wide tree`,
+				before: () => {
+					// Setup
+					tree = generateWideSimpleTree(numberOfNodes, leafValue);
+				},
+				benchmarkFn: () => {
+					writeWideSimpleTreeNewValue(tree, changedLeafValue, tree.values.length - 1);
+				},
+				after: () => {
+					const actual = readWideSimpleTreeValue(tree, tree.values.length - 1);
+					assert.equal(actual, changedLeafValue);
+				},
+			});
+		}
+
+		for (const [numberOfNodes, benchmarkType] of nodesCountWide) {
+			let tree: WideTreeNode;
+			benchmark({
+				type: benchmarkType,
+				title: `Remove and insert first value at leaf of ${numberOfNodes} Wide tree`,
+				before: () => {
+					// Setup
+					tree = generateWideSimpleTree(numberOfNodes, leafValue);
+				},
+				benchmarkFn: () => {
+					writeWideSimpleTreeNewValue(tree, changedLeafValue, 0);
+				},
+				after: () => {
+					const actual = readWideSimpleTreeValue(tree, 0);
+					assert.equal(actual, changedLeafValue);
+				},
+			});
+		}
+
+		for (const [numberOfNodes, benchmarkType] of nodesCountWide) {
+			let tree: WideTreeNode;
+			benchmark({
+				type: benchmarkType,
+				title: `Move second leaf to begining of ${numberOfNodes} Wide tree`,
+				before: () => {
+					// Setup
+					tree = generateWideSimpleTree(numberOfNodes, leafValue);
+					writeWideSimpleTreeNewValue(tree, changedLeafValue, 1);
+				},
+				benchmarkFn: () => {
+					moveWideSimpleTreeValue(tree, 1, 0);
+				},
+				after: () => {
+					const actual = readWideSimpleTreeValue(tree, 0);
+					assert.equal(actual, changedLeafValue);
```

Comment:
THis seems fragile.

If the benchmark runts its timing loop an od number of times, this will fail because you have swapped the first two items an odd number of times.

Currently I think the benchmarking code is unlikely to do this due to how it computes iteration counts, but thats not guaranteed to be the case.

Reply:
I was a little off with my initial take on this. If I understand this correctly, think it will fail if the benchmarkFn is run an even number of times.

The first time the benchmark function is run, it moves the item at index one to index 0, swapping the first two items.

The second time it runs the benchmarkFn, it does the same thing again, swapping them back.

if `after` is run at that point, it would fail, since the item at index 0 would be the item that was originally at index 0, not the change value that was at index 1.

`benchmark` is supposed to be able to run benchmarkFn any number of times (its run as many times as needed to generate accurate timing)

Reply:
I believe the test below has the same issue as well (though its swapping the last two items)

Reply:
So the test case creates an array with size 10, swap 2 items and verify it. Then create an array with size 100, swap 2s and verify it. Do the same for array size 500. Every time it will be a new array and they won't share same memory location. It should be fine...

Reply:
What Craig is pointing out is that the benchmark tool will run the `benchmarkFn` many times for the same test (after running `before()` once, and before running `after()` once), so the same two items will keep swapping over and over. If they were swapped an even number of times, they just end up in their original locations, so the validation in `after()` will fail because it looks like they didn't move at all.

@CraigMacomber got any suggestion on how to validate in this case, or what to do? Off the top of my head, a) initialize a boolean `swappedAtLeastOnce = false` in `before()`, set it to true in `benchmarkFn()` after swapping, check that it's true in `after()`, but that just validates that the function ran, not that the values actually swapped. b) Somehow check if we're in performance mode and do or skip the check in `after()` based on that. That seems a bit overkill, though.
Maybe c), instead of a boolean `swappedAtLeastOnce` use an integer counter, increment it in `benchmarkFn()` after the swap, and in `after()` do the assertion if `counter === 1`, otherwise do nothing?

Reply:
I will vote for skipping the assertion since this is a performance test. We just want to know the execution test for move nodes operation time for wide tree. :/

Reply:
Technically we could, but since these tests run as correctness tests by default (they need the `--perfMode` flag to be run as performance tests), there is value in having them actually validate what they say they do. Otherwise when we run them as performance tests, we might not be actually measuring what we think. I'm leaning towards option c) above, but not sure if we might potentially run into an overflow when the benchmark function is ran a lot of times by the tool.

Reply:
Disabling the validation for all cases exact exactly one iteration isn't great. If you read the docs for `benchmark` which this is using it says `When not {@link isInPerformanceTestingMode}, runs one iteration as a normal Mocha test.`

This making this check conditional on `!isInPerformanceTestingMode` would work.
```suggestion
					// Even number of iterations cancel out, so this validation only works after odd numbers of iterations.
					// Correctness mode always does a single iteration, so just validate that case.
					if (!isInPerformanceTestingMode) assert.equal(tree[0], changedLeafValue);
```